File: ios/Kurir/dependency/auth/AuthModule.swift
//
//  AuthModule.swift
//  Kurir
//
//  Created by Loren on 19/8/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Foundation
import Mobilex

class AuthModule: Module {
    override func inject() {
        // Wrapper
        register(AuthWrapper.self) { r in
            AuthWrapper(repository: r.resolve(AuthRepository.self)!)
        }
        
        // ViewModel
        register(AuthViewModel.self) { r in
            AuthViewModel(r.resolve(AuthWrapper.self)!)
        }
        
        // Delegate
        register(AuthDelegate.self) { (r, view: AuthView) -> AuthDelegate in
            AuthDelegate(view: view, viewModel: r.resolve(AuthViewModel.self)!)
        }
    }
}

File: android/src/main/java/au/com/crazybean/mobilex/kurir/dependency/auth/AuthModule.kt
package au.com.crazybean.mobilex.kurir.dependency.auth

import au.com.crazybean.mobilex.kurir.modules.auth.AuthDelegate
import au.com.crazybean.mobilex.kurir.modules.auth.AuthViewModel
import au.com.crazybean.mobilex.modules.auth.AuthView
import au.com.crazybean.mobilex.modules.auth.AuthWrapper
import org.koin.androidx.viewmodel.dsl.viewModel
import org.koin.dsl.module

val authModule = module {
    // AuthWrapper
    factory {
        AuthWrapper(get())
    }

    // Delegate
    factory { (view: AuthView?) ->
        AuthDelegate(view, get())
    }

    // ViewModel
    viewModel {
        AuthViewModel(get())
    }
}